services/sn_analysis_service/app.py

```diff
@@ -112,74 +134,154 @@ async def run_sn_model(dd):
+                }
+            )
         else:
-            results.update({"status": "failure", "message": "model failed to converge"})
+            log("Fit failed.")
+            rez.update({"status": "failure", "message": "model failed to converge"})
 
     except Exception as e:
-        results.update(
-            {"status": "failure", "message": f"problem running the model {e}"}
+        log(f"Exception while running the model: {e}")
+        log(f"{traceback.format_exc()}")
+        log(f"Data: {data}")
+        rez.update({"status": "failure", "message": f"problem running the model {e}"})
+    finally:
+        # clean up local files
+        for f in local_temp_files:
+            try:
+                os.remove(f)
+            except:  # noqa E722
+                pass
+    return rez
+
+
+class MainHandler(tornado.web.RequestHandler):
+    def set_default_headers(self):
+        self.set_header('Content-Type', 'application/json')
+
+    def error(self, code, message):
+        self.set_status(code)
+        self.write({'message': message})
+
+    def get(self):
+        self.write({'status': 'active'})
+
+    def post(self):
+        """
+        Analysis endpoint which sends the `data_dict` off for
+        processing, returning immediately. The idea here is that
+        the analysis model may take awhile to run so we
+        need async behavior.
+        """
+        try:
+            data_dict = tornado.escape.json_decode(self.request.body)
+        except json.decoder.JSONDecodeError:
+            err = traceback.format_exc()
+            log(f"JSON decode error: {err}")
+            return self.error(400, "Invalid JSON")
+
+        required_keys = ["inputs", "callback_url", "callback_method"]
+        for key in required_keys:
+            if key not in data_dict:
+                log(f"missing required key {key} in data_dict")
+                return self.error(400, f"missing required key {key} in data_dict")
+
+        def sn_analysis_done_callback(
+            future,
+            logger=log,
+            data_dict=data_dict,
+        ):
+            """
+            Callback function for when the sn analysis service is done.
+            Sends back results/errors via the callback_url.
+            """
+            try:
+                result = future.result()
+            except Exception as e:
+                # catch all the exceptions and log them,
+                # try to write back to SkyPortal something
+                # informative.
+                logger(f"{str(future.exception())[:1024]} {e}")
+                result = {
+                    "status": "failure",
+                    "message": f"str(future.exception())[:1024] {e}",
+                }
+            finally:
+                upload_analysis_results(result, data_dict)
+
+        runner = functools.partial(run_sn_model, data_dict)
+        x = IOLoop.current().run_in_executor(None, runner)
```

Comment:
Consider either getting rid of `x` or giving it a more helpful name, like `future_result`.

Reply:
done.